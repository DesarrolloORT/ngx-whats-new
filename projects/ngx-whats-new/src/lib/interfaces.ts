export interface WhatsNewItem {
  title?: string;
  text?: string;
  html?: string;
  image?: {
    src: string;
    bgColor?: string;
    altText: string;
  };
  button?: {
    text: string;
    textColor: string;
    bgColor: string;
    position?: 'left' | 'center' | 'right';
  };
}

export interface DialogOptions {
  disableClose?: boolean;
  enableKeyboardNavigation?: boolean;
  clickableNavigationDots?: boolean;
  customStyle?: {
    maxWidth?: string;
    boxShadow?: string;
    backgroundColor?: string;
    textColor?: string;
    borderRadius?: string;
    borderSize?: string;
  };
}

export interface NavigationEvent {
  previousItem: NavigationItem | undefined;
  currentItem: NavigationItem;
}

interface NavigationItem {
  index: number;
  item: WhatsNewItem;
}
