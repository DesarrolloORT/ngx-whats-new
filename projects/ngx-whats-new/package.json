{
  "name": "ngx-whats-new",
  "version": "1.0.0",
  "homepage": "https://github.com/4gray/ngx-whats-new",
  "author": "4gray",
  "peerDependencies": {
    "@angular/common": "^18.2.8",
    "@angular/core": "^18.2.8",
    "@angular/cdk": "^18.2.10"
  },
  "dependencies": {
    "tslib": "^2.8.0"
  },
  "keywords": [
    "angular",
    "modal",
    "popup",
    "dialog"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/4gray/ngx-whats-new"
  },
  "license": "MIT"
}
