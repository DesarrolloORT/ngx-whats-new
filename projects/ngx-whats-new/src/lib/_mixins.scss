@mixin generate-css-variables($variables-map) {
  @each $name, $value in $variables-map {
    --#{$name}: #{$value};
  }
}
