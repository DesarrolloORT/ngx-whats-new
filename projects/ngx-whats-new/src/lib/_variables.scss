$wn-variables: (
  // Global variables
  wn-base-font-size: 16px,
  wn-base-font-family: "'Segoe UI',Tahoma,Geneva,Verdana,sans-serif",

  // Button focus outline
  wn-button-focus-outline-color: blue,
  wn-button-focus-outline-width: 2px,
  wn-button-focus-outline-style: solid,
  wn-button-focus-outline-offset: 2px,

  // Main button
  wn-main-button-background-color: #222,
  wn-main-button-color: #fff,
  wn-main-button-border: none,
  wn-main-button-padding: 0.7em 2.5em,
  wn-main-button-border-radius: 4px,
  wn-main-button-text-transform: uppercase,
  // hover state
  wn-main-button-hover-opacity: 0.9,
  wn-main-button-hover-color: #fff,
  wn-main-button-hover-background-color: #222,
  // active state
  wn-main-button-active-opacity: 0.9,
  wn-main-button-active-color: #ccc,
  wn-main-button-active-background-color: #444,

  // Close button
  wn-close-button-background-color:
    rgba(
      $color: #222,
      $alpha: 0.5,
    ),
  wn-close-button-color: #fff,
  wn-close-button-border: none,
  wn-close-button-border-radius: 100%,
  wn-close-button-top: 0.5em,
  wn-close-button-right: 0.5em,
  wn-close-button-left: auto,
  wn-close-button-bottom: auto,
  wn-close-button-z-index: 10,

  // Image
  wn-image-background-color: #ddd,

  // Modal container
  wn-modal-container-z-index: 1000,

  // Modal
  wn-modal-background-color: #fff,
  wn-modal-box-shadow: 0 0 1em rgba(0, 0, 0, 0.3),
  wn-modal-border-radius: 4px,

  // Navigation dots
  wn-nav-dot-color: #222,
  wn-nav-dot-background-color: transparent,
  wn-nav-dot-border-radius: 100%,
  wn-nav-dot-border: 1px solid rgba(0, 0, 0, 0.3),
  //hover state
  wn-nav-dot-hover-background-color: rgba(0, 0, 0, 0.1),
  //active state
  wn-nav-dot-active-background-color: #222,
  wn-nav-dot-active-border: 1px solid #222,

  // Title
  wn-title-font-weight: 600,
  wn-title-color: #222,
  wn-title-text-align: center,

  // Text
  wn-text-color: #666,
  wn-text-text-align: center,
  wn-text-font-weight: 400,

  // Link
  wn-text-link-color: #222,
  wn-text-link-text-decoration: underline,
  wn-text-link-hover-color: #222,

  // Backdrop
  wn-backdrop-filter: blur(3px),
  wn-backdrop-z-index: -1,
  wn-backdrop-color: rgba(0, 0, 0, 0.3)
);
