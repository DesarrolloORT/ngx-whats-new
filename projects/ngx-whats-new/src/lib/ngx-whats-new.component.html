@if (isVisible) {
<div class="wn-modal-container">
  <div
    role="dialog"
    class="wn-modal"
    aria-modal="true"
    aria-labelledby="modal-title"
    aria-describedby="modal-description"
    [style.max-width]="options.customStyle?.maxWidth"
    cdkTrapFocus>
    @if (!options.disableClose) {
    <button
      #wnCloseButton
      aria-label="Close dialog"
      class="wn-close-modal-button"
      (click)="close()">
      â›Œ
    </button>
    }
    <div class="wn-modal-layout">
      <div
        class="wn-modal-body"
        aria-live="polite"
        [style.padding-top]="items[selectedIndex].image ? '' : '1.4em'"
        [style.border-radius]="options.customStyle?.borderRadius"
        [style.border]="options.customStyle?.borderSize"
        [style.boxShadow]="options.customStyle?.boxShadow"
        [style.backgroundColor]="options.customStyle?.backgroundColor"
        [style.color]="options.customStyle?.textColor">
        @if (items[selectedIndex].image) { @let img = items[selectedIndex].image;
        <div class="wn-image" [style.background-color]="img?.bgColor">
          <img [src]="img?.src" [alt]="img?.altText" />
        </div>
        }
        <div class="wn-info-body">
          @if (items[selectedIndex].title) {
          <div class="wn-title">{{ items[selectedIndex].title }}</div>
          } @if (items[selectedIndex].text) {
          <div class="wn-text">{{ items[selectedIndex].text }}</div>
          } @if (items[selectedIndex].html) {
          <div class="wn-text" [innerHTML]="items[selectedIndex].html"></div>
          } @if (items[selectedIndex].button) { @let btn = items[selectedIndex].button;
          <button
            class="wn-button"
            [style.align-self]="btn?.position ?? 'center'"
            [style.background-color]="btn?.bgColor"
            [style.color]="btn?.textColor"
            (click)="goToNext()">
            {{ btn?.text }}
          </button>
          }
        </div>
      </div>
      @if (items.length > 1) {
      <nav aria-label="Carousel-like item navigation" class="wn-navigation">
        <ul class="wn-navigation-list" aria-label="Carousel Tabs">
          @for (item of items; track i; let i = $index;) {
          <li
            class="wn-dot"
            role="tabpanel"
            [class.clickable]="options.clickableNavigationDots"
            [class.active]="i === selectedIndex"
            [class.inactive]="i !== selectedIndex">
            <button
              #wnNavButton
              role="tab"
              [attr.aria-selected]="i === selectedIndex"
              [attr.aria-controls]="'tabpanel-' + i"
              [id]="'tab-' + i"
              [tabindex]="i === selectedIndex ? '0' : '-1'"
              (click)="navigateTo(i)"
              [attr.aria-label]="
                (i === selectedIndex ? 'Current slide, ' : '') +
                'Navigate to item ' +
                (i + 1)
              "
              class="wn-nav-button">
              Navigate to item {{ i + 1 }}
            </button>
          </li>
          }
        </ul>
      </nav>
      }
    </div>
  </div>
  <div class="wn-backdrop" aria-hidden="true" (keydown)="close()" (click)="close()"></div>

  <h1 class="sr-only" id="modal-title">Update information dialog</h1>
  <p class="sr-only" id="modal-description">
    This dialog contains information related to the latest features and fixes.
  </p>
</div>
}
