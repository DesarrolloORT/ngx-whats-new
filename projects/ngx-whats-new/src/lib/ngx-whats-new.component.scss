:host {
  position: fixed;
  top: 0;
  left: 0;

  * {
    font-size: 14px;
    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
  }
}

.wn-modal-container {
  position: fixed;
  z-index: 1000;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;

  .wn-modal {
    width: 35em;
    height: auto;
    position: relative;

    button {
      &:focus-visible {
        outline: 2px solid blue;
        outline-offset: 2px;
      }
    }

    .wn-close-modal-button {
      position: absolute;
      top: 0.5em;
      right: 0.15em;
      cursor: pointer;
      background-color: rgba($color: #222, $alpha: 0.5);
      color: #fff;
      border: none;
      aspect-ratio: 1/1;
      border-radius: 100%;
    }

    .wn-modal-layout {
      display: flex;
      flex-direction: column;
      align-items: center;
      gap: 1.4em;
      width: fit-content;

      .wn-modal-body {
        background: #fff;
        overflow: hidden;
        display: flex;
        flex-direction: column;

        .wn-image {
          overflow: hidden;
          align-items: center;
          display: flex;
          justify-content: center;
        }

        .wn-info-body {
          display: flex;
          flex-direction: column;
          align-items: center;
          padding: 1.4em;
          gap: 1.4em;

          .wn-title {
            font-size: 1.3em;
            font-weight: 600;
            color: #222;
            text-align: center;
          }

          .wn-text {
            color: #666;
            text-align: center;
          }

          .wn-button {
            border: none;
            padding: 0.7em 2.5em;
            text-transform: uppercase;
            cursor: pointer;
            outline: none;

            &:hover {
              opacity: 0.9;
            }
          }
        }
      }

      /** Dot navigation **/
      .wn-navigation {
        .wn-navigation-list {
          align-items: center;
          display: flex;
          justify-content: center;
          list-style-type: none;
          padding: 0;
          margin: 0;
          gap: 0.5em;

          .wn-dot {
            border-radius: 100%;
            height: 1em;
            width: 1em;
            position: relative;

            &.clickable {
              &:not(.active) {
                .wn-nav-button {
                  cursor: pointer;
                }

                &:hover {
                  background-color: rgba(0, 0, 0, 0.1);
                }
              }
            }

            .wn-nav-button {
              position: absolute;
              top: 0;
              bottom: 0;
              border-radius: 100%;
              left: 0;
              right: 0;
              width: 100%;
              height: 100%;
              display: block;
              border: none;
              overflow: hidden;
              background-color: transparent;
              color: transparent;
              user-select: none;

              &:focus-visible {
                outline: 2px solid blue;
                outline-offset: 2px;
              }
            }

            &.active {
              background-color: rgba(0, 0, 0, 0.8);
              border: 1px solid rgba(0, 0, 0, 0);
            }

            &.inactive {
              background-color: transparent;
              border: 1px solid rgba(0, 0, 0, 0.3);
            }
          }
        }
      }
    }
  }

  .wn-backdrop {
    backdrop-filter: blur(3px);
    z-index: -1;
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    background-color: rgba(0, 0, 0, 0.3);
  }

  .sr-only {
    position: absolute;
    width: 1px;
    height: 1px;
    padding: 0;
    margin: -1px;
    overflow: hidden;
    clip: rect(0, 0, 0, 0);
    border: 0;
  }
}
