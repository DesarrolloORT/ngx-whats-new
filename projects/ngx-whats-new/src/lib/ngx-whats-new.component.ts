import { A11yModule } from '@angular/cdk/a11y';
import { CommonModule } from '@angular/common';
import {
  AfterViewInit,
  Component,
  ElementRef,
  EventEmitter,
  inject,
  Input,
  OnDestroy,
  Output,
  QueryList,
  Renderer2,
  ViewChild,
  ViewChildren,
} from '@angular/core';

import { DialogOptions, NavigationEvent, WhatsNewItem } from './interfaces';

const DEFAULT_OPTIONS: DialogOptions = {
  customStyle: {
    maxWidth: '500px',
  },
  clickableNavigationDots: true,
  enableKeyboardNavigation: true,
  disableClose: true,
};

@Component({
  selector: 'ngx-whats-new',
  standalone: true,
  imports: [CommonModule, A11yModule],
  templateUrl: './ngx-whats-new.component.html',
  styleUrls: ['./ngx-whats-new.component.scss'],
})
export class NgxWhatsNewComponent implements AfterViewInit, OnDestroy {
  // * -----------------------
  // * Component's public API
  // * -----------------------

  /** Items to show in the dialog */
  @Input() public set items(items: WhatsNewItem[]) {
    if (items && items.length > 0) {
      this._items = items;
    } else {
      console.warn('NgxWhatsNewComponent: No items provided.');
      this._items = [];
    }
    this.resetState();
  }
  public get items(): WhatsNewItem[] {
    return this._items;
  }

  /** DialogOptions for What's New dialog */
  @Input() public set options(options: DialogOptions) {
    this._options = { ...options };
  }
  public get options() {
    return this._options;
  }

  /**
   * @output itemNavigation - Emits an event whenever the selected item changes.
   * The event payload includes the previous and current index and the corresponding items.
   */
  @Output() public readonly navigation = new EventEmitter<NavigationEvent>();

  /** @output closed - Emits an event on dialog close */
  @Output() public readonly closed = new EventEmitter<void>();

  /** @output opened - Emits an event on dialog open */
  @Output() public readonly opened = new EventEmitter<void>();

  /** Navigates to the next item. Closes What's New dialog if it is the last one. */
  public goToNext(): void {
    if (this.items.length === 0) {
      console.warn('NgxWhatsNewComponent: No items to navigate.');
      return;
    }

    if (this.selectedIndex < this.items.length - 1) {
      const previousIndex = this.selectedIndex;
      const previousItem = this.items[previousIndex];
      this.selectedIndex += 1;
      this.updateTabIndices();
      this.emitNavigationEvent(
        previousIndex,
        previousItem,
        this.selectedIndex,
        this.items[this.selectedIndex]
      );
    } else {
      this.close();
    }
  }

  /**
   * Navigates to the item corresponding to the provided number.
   * @param index number of the item
   */
  public navigateTo(index: number): void {
    if (this.items.length === 0) {
      console.warn('NgxWhatsNewComponent: No items to navigate.');
      return;
    }

    if (this._options.clickableNavigationDots && index !== this.selectedIndex) {
      const previousIndex = this.selectedIndex;
      const previousItem = this.items[previousIndex];
      this.selectedIndex = index;
      this.updateTabIndices();
      this.focusButton(index);
      this.emitNavigationEvent(
        previousIndex,
        previousItem,
        this.selectedIndex,
        this.items[this.selectedIndex]
      );
    }
  }

  /** Opens What's New dialog. */
  public open(): void {
    Promise.resolve().then(() => {
      this.isVisible = true;
      this.opened.emit();
      this.registerKeyboardListener();
      this.resetState();
    });
  }

  /** Closes What's New dialog. */
  public close(): void {
    if (!this._options.disableClose) {
      this.unregisterKeyboardListener();
      this.isVisible = false;
      this.closed.emit();
      this.resetState();
    }
  }

  /**
   * Sets focus to the close button when the component is initialized.
   *
   * **MANUAL EXECUTION OF THIS FUNCTION IS NOT RECOMMENDED**
   */
  ngAfterViewInit(): void {
    if (this._closeButton) {
      this._closeButton.nativeElement.focus(); // Set focus to the close button
    }
  }

  ngOnDestroy(): void {
    this.unregisterKeyboardListener();
  }

  // * -------------------------------------------
  // * Private or protected methods and properties
  // * -------------------------------------------

  /** DialogOptions item initialized with default values. Accepts custom values. */
  private _options: DialogOptions = DEFAULT_OPTIONS;

  /** Items to show in the dialog */
  private _items: WhatsNewItem[] = [];

  /** Control whether the dialog is visible */
  protected isVisible = false;

  /** Reference to the close button */
  @ViewChild('wnCloseButton') private readonly _closeButton?: ElementRef;

  /** Reference to the navigation buttons */
  @ViewChildren('wnNavButton')
  private readonly navButtons?: QueryList<ElementRef>;

  /** Index of the selected modal item */
  protected selectedIndex = 0;

  /** Instance of Renderer2 */
  private readonly _renderer: Renderer2 = inject(Renderer2);

  /** Function to register/unregister keyboard listener*/
  private keyboardListener?: () => void;

  /**
   * Resets the component's state to its initial values.
   *
   * This method is called when the dialog is opened or closed to ensure that
   * the component does not retain any state from previous interactions.
   */
  private resetState(): void {
    this.selectedIndex = 0;
    this.updateTabIndices();
  }

  /**
   * Handles keyboard-based navigation between items.
   *
   * Handled events:
   *  - ArrowRight: next item
   *  - ArrowLeft: previous item
   *  - Escape: close dialog
   * @param $event keyboard event
   */
  protected handleKeyboardNavigation($event: KeyboardEvent): void {
    if (this._options.enableKeyboardNavigation) {
      let nextIndex = this.selectedIndex;
      switch ($event.key) {
        case 'ArrowRight':
          if (this.selectedIndex < this.items.length - 1) {
            nextIndex = this.selectedIndex + 1;
          } else {
            this.close();
            return;
          }
          break;
        case 'ArrowLeft':
          if (this.selectedIndex > 0) {
            nextIndex = this.selectedIndex - 1;
          } else return;

          break;
        case 'Escape':
          this.close();
          return;
        default:
          break;
      }

      if (nextIndex !== this.selectedIndex) {
        const previousIndex = this.selectedIndex;
        const previousItem = this.items[previousIndex];
        this.selectedIndex = nextIndex;
        this.updateTabIndices();
        this.focusButton(this.selectedIndex);
        this.emitNavigationEvent(previousIndex, previousItem, nextIndex, this.items[nextIndex]);
        $event.preventDefault();
      }
    }
  }

  /**
   * Updates the `tabindex` and `aria-selected` attributes for all navigation buttons.
   *
   * This method ensures that only the currently selected button is focusable via keyboard navigation
   * by setting its `tabindex` to `0`, while all other buttons have `tabindex` set to `-1`.
   * Additionally, it updates the `aria-selected` attribute to reflect the current selection state,
   * enhancing accessibility for assistive technologies.
   */
  private updateTabIndices(): void {
    this.navButtons?.forEach((ref, index) => {
      const button = ref.nativeElement;
      button.tabIndex = index === this.selectedIndex ? 0 : -1;
      button.setAttribute('aria-selected', index === this.selectedIndex ? 'true' : 'false');
    });
  }

  /**
   * Sets focus to the navigation button at the specified index.
   *
   * This method programmatically focuses the button corresponding to the given index,
   * ensuring that keyboard users receive visual and functional feedback when navigating
   * through carousel items using arrow keys.
   *
   * @param index - The index of the button to focus within the navigation list.
   */
  private focusButton(index: number): void {
    const buttons = this.navButtons?.toArray();

    const button = buttons?.[index].nativeElement;
    if (button) {
      button.focus();
    }
  }

  /**
   * Emits a navigation event with previous and current index and items.
   *
   * Allows external components to react to navigation changes with full context.
   *
   * @param previousIndex - The index before navigation.
   * @param previousItem - The item before navigation.
   * @param currentIndex - The index after navigation.
   * @param currentItem - The item after navigation.
   */
  private emitNavigationEvent(
    previousIndex: number,
    previousItem: WhatsNewItem,
    currentIndex: number,
    currentItem: WhatsNewItem
  ): void {
    if (previousIndex !== currentIndex)
      this.navigation.emit({
        previousItem: {
          index: previousIndex,
          item: previousItem,
        },
        currentItem: {
          index: currentIndex,
          item: currentItem,
        },
      });
  }

  /**
   * Registers keyboard event listener when the dialog opens.
   */
  private registerKeyboardListener(): void {
    this.keyboardListener = this._renderer.listen('window', 'keydown', (event: KeyboardEvent) => {
      this.handleKeyboardNavigation(event);
    });
  }

  /**
   * Unregisters keyboard event listener when the dialog closes.
   */
  private unregisterKeyboardListener(): void {
    if (this.keyboardListener) {
      this.keyboardListener();
      this.keyboardListener = undefined;
    }
  }
}
