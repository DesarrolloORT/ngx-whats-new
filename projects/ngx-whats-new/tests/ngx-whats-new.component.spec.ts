import { A11yModule } from '@angular/cdk/a11y';
import { CommonModule } from '@angular/common';
import { ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
import { By } from '@angular/platform-browser';

import { DialogOptions, WhatsNewItem } from '../src/lib/interfaces';
import { NgxWhatsNewComponent } from '../src/lib/ngx-whats-new.component';

const mockItems: WhatsNewItem[] = [
  {
    title: 'Whats new in v1.0.0',
    html: 'Lorem ipsum dolor sit amet, consectetur adipiscing el aspect et just.<br /><a href="http://google.com">test</a> ',
    image: {
      src: 'https://picsum.photos/500',
      altText: 'In v1.0.0, lorem ipsum dolor sit amet, consectetur adipiscing el aspect et just.',
    },
    button: {
      text: 'Okay',
      textColor: '#fff',
      bgColor: '#333',
      position: 'center',
    },
  },
  {
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing el aspect et just. Spsum dolor sit amet, consectetur adipiscing el aspect et just',
    button: {
      text: 'Got it!',
      textColor: '#fff',
      bgColor: '#333',
      position: 'center',
    },
  },
  {
    image: {
      src: 'https://picsum.photos/500/500',
      altText: 'Lorem ipsum dolor sit amet, consectetur adipiscing el aspect et just.',
    },
    text: 'Very interesting feature',
    button: {
      text: 'Got it',
      textColor: '#fff',
      bgColor: '#333',
      position: 'center',
    },
  },
  {
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing el aspect et just. Spsum dolor sit amet, consectetur adipiscing el aspect et just.',
    image: {
      src: 'https://cdn0.iconfinder.com/data/icons/bakery-10/512/Donut-256.png',
      bgColor: '#333',
      altText: 'Lorem ipsum dolor sit amet, consectetur adipiscing el aspect et just.',
    },
    button: {
      text: 'Lets go',
      position: 'center',
      textColor: '#fff',
      bgColor: '#333',
    },
  },
];

describe('NgxWhatsNewComponent', () => {
  let component: NgxWhatsNewComponent;
  let fixture: ComponentFixture<NgxWhatsNewComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [NgxWhatsNewComponent, CommonModule, A11yModule],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(NgxWhatsNewComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;

    component.items = mockItems;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should use default options if none are provided', () => {
    expect(component.options).toEqual({
      customStyle: {
        maxWidth: '500px',
      },
      clickableNavigationDots: true,
      enableKeyboardNavigation: true,
      disableClose: true,
    });
  });

  it('should set options correctly when provided', () => {
    const customOptions: DialogOptions = {
      customStyle: {
        maxWidth: '600px',
        backgroundColor: '#fff',
      },
      clickableNavigationDots: false,
      enableKeyboardNavigation: false,
      disableClose: false,
    };
    component.options = customOptions;
    expect(component.options).toEqual(customOptions);
  });

  it('should set items correctly when provided', () => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }];
    component.items = items;
    expect(component.items).toEqual(items);
  });

  it('should open the dialog and emit "opened" event', fakeAsync(() => {
    jest.spyOn(component.opened, 'emit');
    component.open();
    tick(); // Resolve the Promise in open()
    fixture.detectChanges();
    expect(component['isVisible']).toBe(true);
    expect(component.opened.emit).toHaveBeenCalled();
  }));

  it('should close the dialog and emit "closed" event when disableClose is false', () => {
    component.options = { disableClose: false };
    component.open();
    jest.spyOn(component.closed, 'emit');
    component.close();
    fixture.detectChanges();
    expect(component['isVisible']).toBe(false);
    expect(component.closed.emit).toHaveBeenCalled();
  });

  it('should navigate to next item when goToNext() is called', () => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }, { title: 'Item 3' }];
    component.items = items;
    component.open();
    jest.spyOn(component.navigation, 'emit');

    expect(component['selectedIndex']).toBe(0);
    component.goToNext();
    expect(component['selectedIndex']).toBe(1);
    expect(component.navigation.emit).toHaveBeenCalledWith({
      previousItem: { index: 0, item: items[0] },
      currentItem: { index: 1, item: items[1] },
    });
  });

  it('should close the dialog when goToNext() is called on the last item', () => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }];
    component.items = items;
    component.options = { disableClose: false };
    component.open();
    jest.spyOn(component, 'close');

    component.goToNext();
    expect(component.close).toHaveBeenCalled();
  });

  it('should navigate to specified index when navigateTo(index) is called', fakeAsync(() => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }, { title: 'Item 3' }];
    component.items = items;
    component.open();
    tick(); // Resolve any pending Promises
    fixture.detectChanges();
    jest.spyOn(component.navigation, 'emit');

    component.navigateTo(2);
    expect(component['selectedIndex']).toBe(2);
    expect(component.navigation.emit).toHaveBeenCalledWith({
      previousItem: { index: 0, item: items[0] },
      currentItem: { index: 2, item: items[2] },
    });
  }));

  it('should not navigate when navigateTo(index) is called and clickableNavigationDots is false', fakeAsync(() => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }];
    component.items = items;
    component.options = { clickableNavigationDots: false };
    component.open();
    tick(); // Resolve any pending Promises
    fixture.detectChanges();
    jest.spyOn(component.navigation, 'emit');

    component.navigateTo(1);
    expect(component['selectedIndex']).toBe(0); // Should not change
    expect(component.navigation.emit).not.toHaveBeenCalled();
  }));

  it('should navigate to next item on ArrowRight keypress', fakeAsync(() => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }];
    component.items = items;
    component.options = { disableClose: false, enableKeyboardNavigation: true };
    component.open();
    tick(); // Resolve any pending Promises
    fixture.detectChanges();
    jest.spyOn(component.navigation, 'emit');

    const event = new KeyboardEvent('keydown', { key: 'ArrowRight' });
    window.dispatchEvent(event);
    fixture.detectChanges();

    expect(component['selectedIndex']).toBe(1);
    expect(component.navigation.emit).toHaveBeenCalledWith({
      previousItem: { index: 0, item: items[0] },
      currentItem: { index: 1, item: items[1] },
    });
  }));

  it('should navigate to previous item on ArrowLeft keypress', fakeAsync(() => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }];
    component.items = items;
    component.options = { disableClose: false, enableKeyboardNavigation: true };
    component.open();
    tick(); // Resolve any pending Promises
    component['selectedIndex'] = 1;
    fixture.detectChanges();
    jest.spyOn(component.navigation, 'emit');

    const event = new KeyboardEvent('keydown', { key: 'ArrowLeft' });
    window.dispatchEvent(event);
    fixture.detectChanges();

    expect(component['selectedIndex']).toBe(0);
    expect(component.navigation.emit).toHaveBeenCalledWith({
      previousItem: { index: 1, item: items[1] },
      currentItem: { index: 0, item: items[0] },
    });
  }));

  it('should close the dialog on Escape keypress when disableClose is false', fakeAsync(() => {
    component.options = { disableClose: false, enableKeyboardNavigation: true };
    component.open();
    tick(); // Resolve any pending Promises
    jest.spyOn(component, 'close');

    const event = new KeyboardEvent('keydown', { key: 'Escape' });
    window.dispatchEvent(event);
    fixture.detectChanges();

    expect(component.close).toHaveBeenCalled();
  }));

  it('should focus on close button after view init when disableClose is false', fakeAsync(() => {
    component.options = { disableClose: false };
    component.items = [{ title: 'Item 1' }];
    component.open();
    tick(); // Resolve any pending Promises
    fixture.detectChanges();

    const closeButton = fixture.debugElement.query(By.css('.wn-close-modal-button'));
    const focusSpy = jest.spyOn(closeButton.nativeElement, 'focus');

    component.ngAfterViewInit();

    expect(focusSpy).toHaveBeenCalled();
  }));

  it('should unregister keyboard listener on ngOnDestroy', () => {
    component.open();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const unregisterSpy = jest.spyOn(component as any, 'unregisterKeyboardListener');

    component.ngOnDestroy();

    expect(unregisterSpy).toHaveBeenCalled();
  });

  it('should close the dialog when ArrowRight is pressed on the last item', fakeAsync(() => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }];
    component.items = items;
    component.options = { disableClose: false, enableKeyboardNavigation: true }; // Allow closing
    component.open();
    tick(); // Resolve the Promise in open()

    jest.spyOn(component, 'close');

    // Set selectedIndex to the last item
    component['selectedIndex'] = items.length - 1;
    fixture.detectChanges();

    // Simulate ArrowRight keypress
    const event = new KeyboardEvent('keydown', { key: 'ArrowRight' });
    window.dispatchEvent(event);
    fixture.detectChanges();

    expect(component.close).toHaveBeenCalled();
  }));

  it('should not navigate when ArrowLeft is pressed on the first item', fakeAsync(() => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }];
    component.items = items;
    component.options = { disableClose: false, enableKeyboardNavigation: true };
    component.open();
    tick();
    fixture.detectChanges();

    jest.spyOn(component.navigation, 'emit');

    // Ensure selectedIndex is 0
    component['selectedIndex'] = 0;

    // Simulate ArrowLeft keypress
    const event = new KeyboardEvent('keydown', { key: 'ArrowLeft' });
    window.dispatchEvent(event);
    fixture.detectChanges();

    // selectedIndex should remain 0
    expect(component['selectedIndex']).toBe(0);
    // Navigation event should not be emitted
    expect(component.navigation.emit).not.toHaveBeenCalled();
  }));

  it('should not navigate or close when an unhandled key is pressed', fakeAsync(() => {
    const items: WhatsNewItem[] = [{ title: 'Item 1' }, { title: 'Item 2' }];
    component.items = items;
    component.options = { disableClose: false, enableKeyboardNavigation: true };
    component.open();
    tick();
    fixture.detectChanges();

    jest.spyOn(component.navigation, 'emit');
    jest.spyOn(component, 'close');

    const initialIndex = component['selectedIndex'];

    // Simulate an unhandled keypress (e.g., 'Enter')
    const event = new KeyboardEvent('keydown', { key: 'Enter' });
    window.dispatchEvent(event);
    fixture.detectChanges();

    expect(component['selectedIndex']).toBe(initialIndex);
    expect(component.navigation.emit).not.toHaveBeenCalled();
    expect(component.close).not.toHaveBeenCalled();
  }));

  it('should warn and not navigate when navigateTo is called with no items', () => {
    jest.spyOn(console, 'warn');

    component.items = [];
    component.navigateTo(0);

    expect(console.warn).toHaveBeenCalledWith('NgxWhatsNewComponent: No items to navigate.');
    expect(component['selectedIndex']).toBe(0);
  });

  it('should warn and not navigate when goToNext is called with no items', () => {
    jest.spyOn(console, 'warn');

    component.items = [];
    component.goToNext();

    expect(console.warn).toHaveBeenCalledWith('NgxWhatsNewComponent: No items to navigate.');
    expect(component['selectedIndex']).toBe(0);
  });

  it('should warn and set items to empty array when items is null', () => {
    jest.spyOn(console, 'warn');

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (component.items as any) = null;

    expect(console.warn).toHaveBeenCalledWith('NgxWhatsNewComponent: No items provided.');
    expect(component.items).toEqual([]);
  });

  it('should warn and set items to empty array when items is empty', () => {
    jest.spyOn(console, 'warn');

    component.items = [];

    expect(console.warn).toHaveBeenCalledWith('NgxWhatsNewComponent: No items provided.');
    expect(component.items).toEqual([]);
  });
});
