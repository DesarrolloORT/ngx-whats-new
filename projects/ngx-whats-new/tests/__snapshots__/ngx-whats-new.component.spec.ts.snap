// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NgxWhatsNewComponent should match snapshot 1`] = `
<div
  id="root1"
>
  
</div>
`;
